### ISSUE TITLE
IDE should highlight required members when in an initializer

### ISSUE THREAD

**Issue by jasonmalinowski** [dotnet/roslyn#61534]
C# 11 is bringing [required members](https://github.com/dotnet/csharplang/blob/main/proposals/required-members.md), where certain fields/properties can be marked as they _must_ be initialized in an object initializer. @RikkiGibson had the good suggestion that we should highlight these in some way in completion so these properties stand out from the properties that aren't required. Right now we will show the "required" keyword in the tooltip if you select (but not commit) a required member in completion, but that's not very useful.

There's a few obvious options, and these aren't mutually exclusive:

1. Put a "required" modifier directly in the text, or in the gray text that's to the right that we normally use for the namespace for unimported members.
2. Use a different icon.
3. Make the required entries bold/italics/[insert text decoration here]
4. Sort the required members to the top.

My _preference_ is the first option and maybe also the fourth. My worry is an icon would require a completely different icon to be easily seen -- a small modifier/overlay might be hard to see in a list of members. Bold/italic members might be hard to read if you have camel-case matching already there, and users won't have an easy way to discover what the bold/italics actually mean. Just saying "required" as a text marker gives them something to start searching from, and also connects it to the modifier directly.

**Comment by jasonmalinowski** [comment]
> I imagine (4) could be helpful because users would just be able to hit enter when the completion list pops up with a required member at the top.

Ah, that's a good subtlety here: we can also leave the sorting as is, but just change the preselection logic which we can control independently of the sort. So we actually have two options here.

**Comment by CyrusNajmabadi** [comment]
i agree that required members should show up above the rest :)

**Comment by davidwengier** [comment]
I like the idea of sorting and an icon, as it mirrors what happens with IntelliCode, so there is precedence for the pattern. I also wonder if sorting without an icon could confuse people, precisely because of the IntelliCode behaviour potentially now being considered the norm.

**Comment by jasonmalinowski** [comment]
> I could also imagine an experience where typing the { in new Widget() { opens ups an initializer block that pre-fills Prop = , for all the required properties, but I think that would end up being daunting/disorienting, as the user will have to figure out how to position their cursor to provide each property value.

@RikkiGibson: the snippet experience we'll track in https://github.com/dotnet/roslyn/issues/61550. To answer your question though about positioning, we already have precedent for tab jumping between fields in snippets, which also works like our argument completion feature too.

**Comment by jasonmalinowski** [comment]
> I also wonder if sorting without an icon could confuse people

Yeah, I wasn't a huge fan of just sort and _nothing_ else, since it might get confusing. Or it looks like a bug if you just have two parameters.

It also occurred to me that we might want to make pre-select be smart so that it picks one of the required members that aren't already provided, so each time it offers the next one. Bonus for keeping things in either alphabetical or source order.

**Comment by jasonmalinowski** [comment]
We decided to make the two following changes:

1. We'll use the gray text for an entry to say "required". There was some concern what we do if we're also showing member types there down the road, but saying 'required int' seems an entirely reasonable thing to do there so that seemed to not be boxing us out of problems.
2. We will adjust preselect so invoking the list without typing a character will prefer a required member over any other required member. We'd choose a required member that hasn't already been specified if possible. What we'll do as a tiebreaker isn't entirely clear.

We decided these options are things to pursue later, if we the above don't work:

1. Changing sorting -- we don't really do that today; if we also wanted to generally introduce a "we sort by importance" or "we sort by inheritance hierarchy" (random examples) down the road then we'd be having this specific sorting behavior colliding with that in potentially confusing ways. 
2. Filters/tabs -- absolutely doable, just requires the engineering work to do it. But since it requires a certain amount of plumbing, we'll wait towards feedback. We thought such an experience would be most useful only if you had fairly large member lists with also a lot of required members as well that you needed to do such filtering, but we might need a bit more info on that scenario first.
3. Bolding/adorning the text: bold we realized was a problem because we currently use bold to represent camel case matching, and we are also thinking about using for more general things like static/inherited/etc. Not sure it made sense to use as general purpose of a mechanism for such a more specific scenario here.

Icons were generally disliked as an option, since that potentially also raises accessibility challenges.
